File: StriverA2Z/C++/6_LinkedList/Questions/DoublyLinkedList/reverse_dll.cpp
#include<iostream>
#include "Node.h"

using namespace std;

Node *reverseDLL(Node *head) {
    if (head == nullptr || head->next == nullptr) return head;
    Node *temp = head;
    Node *prevNode = nullptr;
    while (temp) {
        prevNode = temp->prev;
        temp->prev = temp->next;
        temp->next = prevNode;
        temp = temp->prev;
    }
    return prevNode->prev;
}

int main() {
    Node *head = arrayToLinkedList({12, 8, 5, 7});
    traverseToLinkedList(head);
    head = reverseDLL(head);
    traverseToLinkedList(head);
    return 0;
}

File: StriverA2Z/Java/LinkedLists/Questions/DoublyLinkedList/ReverseDLL.java
public class ReverseDLL {
    public static void main(String[] args) {
        int[] arr = {1,2,3,4,5};
        Node head = LinkedListOperations.arrayToLinkedList(arr);
        head = reverseDLL(head);
        LinkedListOperations.traverseLinkedList(head);
    }

    static Node reverseDLL(Node head) {
        if (head == null || head.next == null) return head;
        Node temp = head;
        Node prevNode = null;
        while (temp != null) {
            prevNode = temp.prev;
            temp.prev = temp.next;
            temp.next = prevNode;
            temp = temp.prev;
        }
        return prevNode.prev;
    }
}
